code/services/QueuedJobService.php

```diff
@@ -388,8 +401,20 @@ public function runJob($jobId) {
 			} else {
 				$jobDescriptor->JobRestarted = date('Y-m-d H:i:s');
 			}
-			
-			$jobDescriptor->JobStatus = QueuedJob::STATUS_RUN;
+
+			// write the status and determine if any rows were affected, for protection against a
+			// potential race condition where two or more processes start the same job at once.
+			// This deliberately does not use write() as that would always update LastEdited
+			// and thus the row would always be affected.
+			DB::query(sprintf(
```

Comment:
I don't think we need to modify this code here. The only way for us to actually "grab" the job is to go through the `initialiseJob`, which already has the race condition prevention. There is another potential source of problems down below though: there is a `write` in the "catch" block - if the "second" job encounters an exception in the `initialiseJob`, it will rewrite the status as STATUS_BROKEN, even though the "first" job might be merrily running along.

I think I'd suggest to refactor this:
- separate the UPDATE (the "grab" code) out into a separate function, and call it "grabJob" or something like that.
- pull it out of the try block. This way if the process fails to "grab" the job, it will just exit, instead of trying to do writes.